### ISSUE TITLE
Pulumi Dropping Resources for no reason

### ISSUE THREAD

**Issue by joebeernink** [pulumi/pulumi#16298]
### What happened?

We've been working just fine with no major changes to our infrastructure or Pulumi versions, and then this afternoon, Pulumi started attempting to drop random databases, role assignments and keyvault secrets in multiple pipelines in multiple environments. Pulumi Preview will say nothing about the environment is changing. Pulumi refresh will show nothing is changing. And then Pulumi up goes nuts and start deleting things.

We're running our Pulumi steps in GitHub actions, with Azure Blob storage as our backing store.

We were running CLI version 3.103.1, and tried going to 3.117.0 to fix the problem

We tried running deployments that worked fine yesterday, and now result in resources getting dropped.

This has brought down multiple production instances with no way to fail back. Help needed urgently.

### Example

Not sure how to repo this.

### Output of `pulumi about`

Run pulumi/actions@v5
  with:
    command: up
    work-dir: Applications/<redacted>
    stack-name: prd-eus[2]<redacted>
    secrets-provider: azurekeyvault://<redacted>/keys/pulumi
    cloud-url: azblob://pulumi/twin/infra?storage_account=<redacted>
    refresh: true
    comment-on-pr: true
    upsert: true
    diff: true
    pulumi-version: [3](https://github.com/<redacted>:8:3).103.1
    comment-on-summary: false
    github-token: ***
    expect-no-changes: false
    target-dependents: false
    remove: false
    edit-pr-comment: true
    color: auto
    exclude-protected: false
    suppress-outputs: false
    suppress-progress: false
  env:
    DOTNET_ROOT: /usr/share/dotnet
    AZURE_KEYVAULT_AUTH_VIA_CLI: true
    ARM_USE_OIDC: true
    ARM_CLIENT_ID: ***
    ARM_TENANT_ID: <redacted>
    ARM_CLIENT_SECRET: ***
    AZURE_B2C_TENANT_ID: 
    AZURE_B2C_APP_MANAGER_CLIENT_ID: <redacted>
    AZURE_B2C_APP_MANAGER_CLIENT_SECRET: ***
    IS_PRODUCTION: true
Configured range: 3.103.1
/usr/local/bin/pulumi version
v3.117.0
Pulumi 3.117.0 does not satisfy the desired version 3.103.1. Proceeding to download
Matched version: v3.103.1
Install destination is /home/runner/.pulumi
Successfully deleted pre-existing /home/runner/.pulumi/bin
/usr/bin/tar xz --warning=no-unknown-keyword --overwrite -C /home/runner/.pulumi -f /home/runner/work/_temp/d2c387f[5]mi/bin
/usr/bin/tar xz --warning=no-unknown-keyword --overwrite -C /home/runner/.pulumi -f /home/runner/work/_temp/d2c387f[5](https://github.com/WillowInc/Infrastructure-and-Application-Deployment/actions/runs/9293206863/job/25627729122#step:8:5)-8135-4835-a90b-29d655519c08

### Contributing

Vote on this issue by adding a 👍 reaction. 
To contribute a fix for this issue, leave a comment (and link to your pull request, if you've opened one already).

**Comment by joebeernink** [comment]
What is really odd is that it isn't even dropping everything in a stack. For instance, a managed identity had two roles granted in a keyvault: KeyVaultSecretsUser and KeyVault Reader. It dropped the KeyVaultSecrets user, but left the reader there.

**Comment by joebeernink** [comment]
We have an explanation for why it's not deleting everything... we protect some of our resources, so when it hits one of those, it stops deleting everything. But if you run it again, it deletes a little more and a little more. So it looks like the generated Pulumi stack on Pulumi up is completely empty (but it shouldn't be). I suspect something is throwing an error somewhere, and instead of stopping the process, it's being caught and swallowed.

**Comment by joebeernink** [comment]
We found the root cause. A step in our deployments referred to a DNS entry that had been changed to a wrong url, causing the pipeline to fail unexpectedly. The call was only made in Pulumi Up, not in Preview. I would have expected the whole job to stop at that point, but it instead it proceeded with an empty stack. We'll look at our error handling and try to figure out why this fell through the way it did. Hopefully others can learn from our mistake.

As the saying goes... it's always DNS.
